File: class.lua
---Lote Antecessor comum---
Lote={}
Lote.__index=Lote

function Lote.new(id,owner,categoria)
	local self=setmetatable({},Lote)
	self.id=id
	self.owner=owner
	self.categoria=categoria
	return self
end

function Lote:Preco()
	return 0
end

function Lote:Id()
	return self.id
end

function Lote:Owner()
	return self.owner
end

function Lote:Area()
	return 0
end

function Lote:Info()
	return ""
end

--- Apto ---
Apto={};
Apto.__index=Apto
function Apto.new(id, owner, categoria, nQuartos, nVagas, andar, area_construida, pm2_construida, lazer, nAndares)
	local self=setmetatable({},Apto)
	self.id=id
	self.owner=owner
	self.categoria=categoria
	self.nQuartos=nQuartos
	self.nVagas=nVagas
	self.andar=andar
	self.area_construida=area_construida
	self.pm2_construida=pm2_construida
	self.lazer=lazer
	self.nAndares=nAndares
	return self
end

--Heranca
setmetatable(Apto,{__index = Lote})

--overrides
function Apto:Preco()
	local p=self.pm2_construida * self.area_construida * (0.9 + self.andar/self.nAndares)
	if self.lazer == 'S' then
		return p*1.15
	else
		return p
end

function Apto:Area()
	return self.area_construida
end

--- Terreno ---

File: lp4/src/lp4/class.clj
(in-ns 'lp4.main)

(defprotocol Lote
	(Id [this])
	(Owner [this])
	(Categoria [this])
	(Preco [this])
	(Area [this])
	(Info [this]))

(defrecord Terreno[id owner categoria solo pm2 base1 base2 altura]
	Lote
	(Id [this]
		(:id this))
	(Owner [this]
		(:owner this))
	(Categoria [this]
		(:categoria this))
	(Area [this]
		(case (:categoria this)
			"retang" (* (:base1 this) (:altura this))
			"triang" (/ (* (:base1 this) (:altura this)) 2)
			"trapez" (* (/ (+ (:base1 this) (:base2 this)) 2) (:altura this))
			0))
	(Preco [this]
		(case (:solo this)
			"A" (* (* (:pm2 this) (Area this)) 0.9)
			"R" (* (* (:pm2 this) (Area this)) 1.1)
			"G" (* (* (:pm2 this) (Area this)) 1.3)
			0))
	(Info [this]
		(:solo this))
)

(defrecord Casa[id owner nquartos nvagas npav apav pm2pav al pm2al]
	Lote
	(Id [this]
		(:id this))
	(Owner [this]
		(:owner this))
	(Categoria [this]
		"casa")
	(Area [this]
		(* (:npav this) (:apav this)))
	(Preco [this]
		(+ (* (Area this) (:pm2pav this)) (* (:pm2al this) (:al this))))
	(Info [this]
		(:nquartos this))
)

(defrecord Apto[id owner nquartos nvagas andar ac pm2ac lazer andares]
	Lote
	(Id [this]
		(:id this))
	(Owner [this]
		(:owner this))
	(Categoria [this]
		"apto")
	(Area [this]
		(:ac this))
	(Preco [this]
		(case (:lazer this)
			"N" (* (:pm2ac this) (* (:ac this) (+ 0.9 (/ (:andar this) (:andares this)))))
			"S" (* (* (:pm2ac this) (* (:ac this) (+ 0.9 (/ (:andar this) (:andares this))))) 1.15)
			0))

	(Info [this]
		nil)
)